### PR TITLE
Adds AuthenticationMethodRetriever.

### LINKED ISSUE #78: Change registry protocol default to HTTPS
Remote registries should send over HTTPS. However, local registries may not have HTTPS enabled. 

For local registries, an 'allow_http' flag could allow the use of HTTP protocol.

This is in regards to #68, where the default protocol used for manifest pulling is HTTP.

### LINKED ISSUE #94: Add registry authenticator for other registries.
In `RegistryAuthenticators`, there is currently `RegistryAuthenticator#forDockerHub`.

Add `RegistryAuthenticator#forOther` that:

1) Fetches GET /v2/
2) Extracts the WWW-Authenticate header (eg. `WWW-Authenticate: Bearer realm="https://gcr.io/v2/token",service="gcr.io"`)
3) Parses the header and authenticates accordingly.

### PR DESCRIPTION
**Pull request by coollog**
Step 6 of #59
  
Fixes #78

Fixes #94

### CODE REVIEW

File: crepecake/src/main/java/com/google/cloud/tools/crepecake/registry/RegistryClient.java [hunk 1]
```diff
@@ -50,6 +51,42 @@ public RegistryClient(@Nullable Authorization authorization, String serverUrl, S
...
+  /** Gets the {@link RegistryAuthenticator} to authenticate pulls from the registry. */
+  public RegistryAuthenticator getRegistryAuthenticator()
+      throws IOException, RegistryException, RegistryAuthenticationFailedException {
+    // Gets the WWW-Authenticate header (eg. 'WWW-Authenticate: Bearer realm="https://gcr.io/v2/token",service="gcr.io"')
+    AuthenticationMethodRetriever authenticationMethodRetriever =
+        new AuthenticationMethodRetriever(registryEndpointProperties);
+
+    try {
+      callRegistryEndpoint(authenticationMethodRetriever);
+      throw new RegistryErrorExceptionBuilder(authenticationMethodRetriever.getActionDescription())
```
Comment (loosebazooka) [inline crepecake/src/main/java/com/google/cloud/tools/crepecake/registry/RegistryClient.java]: It seems weird for us to have code that is expecting an exception?

File: crepecake/src/main/java/com/google/cloud/tools/crepecake/registry/AuthenticationMethodRetriever.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  @Override
+  public List<String> getAccept() {
+    return Collections.emptyList();
+  }
+
+  @Override
+  public RegistryAuthenticator handleResponse(Response response) throws RegistryErrorException {
+    throw new RegistryErrorExceptionBuilder(getActionDescription())
```
Comment (loosebazooka) [inline crepecake/src/main/java/com/google/cloud/tools/crepecake/registry/AuthenticationMethodRetriever.java]: Maybe I'm missing something, but this seems to be always throwing this exception?

Comment (coollog) [reply crepecake/src/main/java/com/google/cloud/tools/crepecake/registry/AuthenticationMethodRetriever.java]: Yea, the RegistryAuthenticator expects a `401 Unauthorized` HTTP error, so if it receives a 200 status response (handled by `handleResponse`), it throws an exception.

Comment (loosebazooka) [reply crepecake/src/main/java/com/google/cloud/tools/crepecake/registry/AuthenticationMethodRetriever.java]: Oh I see, this just picks up the authentication method... is this really the only way to do it? Isn't there an endpoint to hit to ask?

Comment (coollog) [reply crepecake/src/main/java/com/google/cloud/tools/crepecake/registry/AuthenticationMethodRetriever.java]: Yeah, this is doing the authentication according to https://docs.docker.com/registry/spec/auth/token/, where:

1) Get the authentication method from the WWW-Authenticate header
2) Authenticate with the provided method (done by `RegistryAuthenticator`)

Comment (loosebazooka) [reply crepecake/src/main/java/com/google/cloud/tools/crepecake/registry/AuthenticationMethodRetriever.java]: Damn that's so silly. okay :smile:

Comment (coollog) [reply crepecake/src/main/java/com/google/cloud/tools/crepecake/registry/AuthenticationMethodRetriever.java]: Here, we are doing it once before all pull calls. But, we might want to have this be done every time we receive a 401 Unauthorized, so I filed https://github.com/GoogleCloudPlatform/minikube-build-tools-for-java/issues/96

Comment (loosebazooka) [reply crepecake/src/main/java/com/google/cloud/tools/crepecake/registry/AuthenticationMethodRetriever.java]: Not sure if you linked to that somewhere, but it'd be nice to include in the code so someone looking at this in the future will understand.

Comment (loosebazooka) [reply crepecake/src/main/java/com/google/cloud/tools/crepecake/registry/AuthenticationMethodRetriever.java]: just kdiding I found it in the code...

Comment (coollog) [reply crepecake/src/main/java/com/google/cloud/tools/crepecake/registry/AuthenticationMethodRetriever.java]: Yeah, it is linked in `RegistryAuthenticator`. `AuthenticationMethodRetriever` just does standard `WWW-Authenticate` header retrieval.

**Pull request merged**